lib/puma/thread_pool.rb

```diff
@@ -155,7 +155,15 @@ def <<(work)
 
     def wait_until_not_full
       @mutex.synchronize do
-        until @todo.size - @waiting < @max - @spawned or @shutdown
+        while true
+          return if @shutdown
+          return if @waiting > 0
+
+          # If we can still spin up new threads and there
+          # is work queued, then accept more work until we would
+          # spin up the max number of threads.
+          return if @todo.size < @max - @spawned
```

Comment:
This is not an equivalent of initial condition. With this change puma serving a single-threaded app with "threads 1,1" attempts to pass a request to a worker that is already serving a request. We're seeing occasional deadlocks for the requests coming right after a long-running request  in our environment with "threads 1,1", "queue_requests false", and "workers 5" configuration. Rolling back this single change to the original condition solves the issue completely.

Reply:
In the case of a single-threaded worker, `@waiting` will be 0, `@todo.size` will be 1, `@max` will be 1 and `@spawned` will be 1, so the condition evaluates to `1 < 1 - 1`, which is zero, so Puma waits here.

Reply:
The problem is in line:
`+          return if @waiting > 0`

This is not how the original condition worked: `until @todo.size - @waiting < @max - @spawned or @shutdown`. The fact that `@waiting > 0` is not enough for decision not to do `@not_full.wait @mutex`.

I prepared a simple test scenario which demonstrates the problem: https://gist.github.com/ivanpesin/f40a76e51a60a83f1190113fdecddaa5

With this commit, second request gets *always blocked* until first request is served, even though there are available workers. Third requests goes to available worker.

With original condition, second request goes to next available worker and works as expected. The condition needs to be either reverted to original form or needs logic correction.

Reply:
We also had this issue.

Our setup is using 16 workers with 1 thread each running with an nginx server in front. We were getting high wait times between nginx and rails as measured by a middleware similar to the one at https://github.com/puma/puma/issues/1405#issuecomment-330806817. We partially undid the change in this pull request with the patch below and our wait times dropped as can be seen by the attached screenshot.

````ruby
module PumaMonkeypatch
  def wait_until_not_full
    @mutex.synchronize do
      until @todo.size - @waiting < @max - @spawned or @shutdown
        @not_full.wait @mutex
      end
    end
  end
end
Puma::ThreadPool.prepend(::PumaMonkeypatch)
````

![image](https://user-images.githubusercontent.com/2487833/33135788-74b353fa-cf71-11e7-88a3-093b918df7f4.png)

Thoughts on how to proceed? Should we file a new issue?

Reply:
> Should we file a new issue?

Please do.

Reply:
Thanks. Ticket created: https://github.com/puma/puma/issues/1471